### PR TITLE
Extended v3io implementation

### PR DESCRIPTION
**Pull request by hayesgb**
- Added private function (_details) to extend compatibility with fsspec
- Extended fsspec compatibility to v>= 0.6.2
- Added error handling to ls() method when container has no prefixes, and when container has no objects.
- Fixed error in formatting container/user paths
- Added handling case for ls() when v3io sdk returns 404 error
- updated info() method from fsspec
- updated find() method from fsspec (this may not be necessary - additional work needed)

### CODE REVIEW

File: v3iofs/fs.py [hunk 1]
```diff
@@ -12,17 +12,23 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
+import logging
```
Comment (tebeka) [inline v3iofs/fs.py:16]: Please group stdlib import togegher (no empty line)

File: v3iofs/fs.py [hunk 2]
```diff
@@ -52,29 +58,79 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+            pathlist.append(data)
+        return pathlist
+
     def ls(self, path, detail=True, **kwargs):
         """Lists files & directories under path"""
+        
         container, path = split_container(path)
-
-        if not container:
+        if container == "":
```
Comment (tebeka) [inline v3iofs/fs.py:79]: Why?

File: v3iofs/fs.py [hunk 3]
```diff
@@ -52,29 +58,79 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+                path=path,
+                get_all_attributes=True,
+                raise_for_status=[HTTPStatus.OK],
+            )
+
+            # Try to fetch a list of directories, else return empty
+            try:
+                prefixes = resp.output.common_prefixes
+                dirs = [prefix_info(container, p) for p in prefixes]
+            except:
```
Comment (tebeka) [inline v3iofs/fs.py:98]: Don't use bare `except`

File: v3iofs/fs.py [hunk 4]
```diff
@@ -12,17 +12,23 @@
...
+import logging
+
 from datetime import datetime, timezone
 from http import HTTPStatus
 from os import environ
 from urllib.parse import urlparse
 
 from fsspec.spec import AbstractFileSystem
 from v3io.dataplane import Client
+import sys
```
Comment (tebeka) [inline v3iofs/fs.py:25]: This is unused import. Please run `make test` to check that all sources pass `flake8`

File: v3iofs/fs.py [hunk 5]
```diff
@@ -52,29 +58,79 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
-        prefixes = resp.output.common_prefixes  # directories
-        fn = prefix_info if detail else prefix_path
-        dirs = [fn(container, p) for p in prefixes]
-
-        objects = resp.output.contents  # files
-        fn = object_info if detail else object_path
-        files = [fn(container, o) for o in objects]
-
-        return dirs + files
+        elif container and path == "":
```
Comment (tebeka) [inline v3iofs/fs.py:81]: No need for `elif` after a `return`

File: v3iofs/fs.py [hunk 6]
```diff
@@ -52,29 +58,79 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+            containers = self._list_containers(detail=True)
+            containers = [c['name'] for c in containers]
+            if not container in containers:
```
Comment (tebeka) [inline v3iofs/fs.py:84]: `if container not in containers:`

File: v3iofs/fs.py [hunk 7]
```diff
@@ -52,29 +58,79 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+
+            # Try to fetch a list of directories, else return empty
+            try:
+                prefixes = resp.output.common_prefixes
+                dirs = [prefix_info(container, p) for p in prefixes]
+            except:
+                dirs = []
+
+            # Try to fetch a list of files, else return empty
+            try:
```
Comment (tebeka) [inline v3iofs/fs.py:102]: Don't have nested `try` unless you really have to. Refactor.

File: v3iofs/fs.py [hunk 8]
```diff
@@ -52,29 +58,79 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+                objects = resp.output.contents
+                files = [object_info(container, o) for o in objects]
+            except:
+                files = []
```
Comment (tebeka) [inline v3iofs/fs.py:106]: You are masking an error here. Don't do that, see the Python zen

File: v3iofs/fs.py [hunk 9]
```diff
@@ -52,29 +58,79 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+            pathlist = dirs + files
+            if not pathlist:
+                raise FileNotFoundError(f"Nothing found in {path}")
+        except:
+            logger.debug("Nothing found in container.  Moving to top level container")
```
Comment (tebeka) [inline v3iofs/fs.py:111]: I'm not sure this is the logic we want. Also too much code inside an `except` clause.

File: v3iofs/fs.py [hunk 10]
```diff
@@ -52,29 +58,79 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+            dirname, _, filename = path.rpartition("/")
+            resp = self._client.get_container_contents(
+                container=container,
+                path=dirname,
+                )
+            objects = resp.output.contents
+            fn = object_info if detail else object_path
+            tempfiles = [object_info(container, o) for o in objects]
+            fullpath = f"/{container}/{path}"
+            logger.debug(f"fullpath:  {fullpath}")
```
Comment (tebeka) [inline v3iofs/fs.py:122]: Don't construct strings in log messages. Use `%s' and pass arguments.

```python
logger.debug('fullpath: %s', fullpath)
```

File: v3iofs/fs.py [hunk 11]
```diff
@@ -52,29 +58,79 @@ def __init__(self, v3io_api=None, v3io_access_key=None, **kw):
...
+            for f in tempfiles:
+                if f['name'] == fullpath:
+                    pathlist.append(f)
+            if not pathlist:
+                raise FileNotFoundError("File or directory not found!!")
+        pathlist = self._details(pathlist)
+        if detail:
+            return pathlist
+        else:
```
Comment (tebeka) [inline v3iofs/fs.py:131]: No need for `else` after return

File: v3iofs/fs.py [hunk 12]
```diff
@@ -106,6 +162,64 @@ def touch(self, path, truncate=True, **kwargs):
...
 
+    def info(self, path, **kwargs):
+        """Give details of entry at path
+        Returns a single dictionary, with exactly the same information as ``ls``
+        would with ``detail=True``.
+        The default implementation should calls ls and could be overridden by a
+        shortcut. kwargs are passed on to ```ls()``.
+        Some file systems might not be able to measure the file's size, in
+        which case, the returned dict will include ``'size': None``.
+        Returns
```
Comment (tebeka) [inline v3iofs/fs.py:173]: You need an empty line before `Returns`, see [here](https://numpydoc.readthedocs.io/en/latest/format.html)

File: v3iofs/fs.py [hunk 13]
```diff
@@ -106,6 +162,64 @@ def touch(self, path, truncate=True, **kwargs):
...
+        -------
+        dict with keys: name (full path in the FS), size (in bytes), type (file,
+        directory, or something else) and other FS-specific keys.
+        """
+
+        out = self.ls(path, detail=True, **kwargs)
+        path = path.rstrip("/")
+        out1 = [o for o in out if o["name"].rstrip("/") == path]
```
Comment (tebeka) [inline v3iofs/fs.py:181]: Find a better name than `out1`

File: v3iofs/fs.py [hunk 14]
```diff
@@ -106,6 +162,64 @@ def touch(self, path, truncate=True, **kwargs):
...
+                files.update(dirs)
+            out.update({info["name"]: info for name, info in files.items()})
+        if self.isfile(path) and path not in out:
+            # walk works on directories, but find should also return [path]
+            # when path happens to be a file
+            out[path] = {}
+        names = sorted(out)
+        if not detail:
+            return names
+        else:
```
Comment (tebeka) [inline v3iofs/fs.py:219]: No `else` after return

File: v3iofs/fs.py [hunk 15]
```diff
@@ -155,12 +268,6 @@ def info_of(container_name, obj, name_key):
     return {
         'name': prefix_path(container_name, getattr(obj, name_key)),
         'size': getattr(obj, 'size', None),
-        'created': parse_time(obj.creating_time),
```
Comment (tebeka) [inline v3iofs/fs.py:158]: Why?

File: v3iofs/fs.py (hunk 1 above)
Comment (hayesgb) [reply v3iofs/fs.py:16]: Linted

File: v3iofs/fs.py (hunk 2 above)
Comment (hayesgb) [reply v3iofs/fs.py:79]: If passing fs.ls("/"), expect to return a list of top level directories.

File: v3iofs/fs.py (hunk 3 above)
Comment (hayesgb) [reply v3iofs/fs.py:99]: Resolved with linting.

File: v3iofs/fs.py (hunk 7 above)
Comment (hayesgb) [reply v3iofs/fs.py:103]: Original master specified raise_for_status=HTTP.OK from v3io dataplane response object, which returns a RuntimeError when a) get_container_contents() returns no response object, but also creates a RuntimeError if needed downstream attributes are not present (response.output.common_prefixes or response.output.contents).  Changed this to check for the presence of these attributes and handle.

File: v3iofs/fs.py (hunk 13 above)
Comment (hayesgb) [reply v3iofs/fs.py:179]: Updated to pathlist

File: v3iofs/fs.py (hunk 6 above)
Comment (hayesgb) [reply v3iofs/fs.py]: obj.creating_time was returning an AttributeError, and that attribute is not needed by fsspec

**Comment by tebeka** [comment]
Thanks @hayesgb . I'll have a look (my) tomorrow (GMT + 3)

File: .gitignore [hunk 16]
```diff
@@ -1,3 +1,4 @@
+*notebooks/*
```
Comment (tebeka) [inline .gitignore:1]: What is this directory? Is it missing from git or autogenerated?

File: setup.py [hunk 17]
```diff
@@ -2,11 +2,11 @@
 
 
 def version():
-    with open('v3iofs/__init__.py') as fp:
+    with open("v3iofs/__init__.py") as fp:
```
Comment (tebeka) [inline setup.py:5]: Please refrain from such changes (`'` to `"`), it's just white noise. Try to have a little as possible in the PR

File: v3iofs/fs.py (hunk 2 above)
Comment (tebeka) [reply v3iofs/fs.py:75]: I was asking about the change from `if not container` to `if contanier != ""`

File: v3iofs/fs.py [hunk 18]
```diff
@@ -37,49 +38,97 @@ class V3ioFS(AbstractFileSystem):
...
-
-        return dirs + files
+                container=container,
+                path=path,
+                get_all_attributes=True,
+                raise_for_status="never",
+            )
+
+        # Try to fetch a list of directories, else return empty
+        if hasattr(resp.output, 'common_prefixes'):
```
Comment (tebeka) [inline v3iofs/fs.py:93]: Extract to a `_resp_dirs` utility function, something like:

```python
def _resp_dirs(resp, container):
    if not hasattr(resp.output, 'common_prefixes'):
        return []

    prefixes = resp.output.common_prefixes
    return [prefix_info(container, p) for p in prefixes]
```

File: v3iofs/fs.py [hunk 19]
```diff
@@ -37,49 +38,97 @@ class V3ioFS(AbstractFileSystem):
...
+            prefixes = resp.output.common_prefixes
+            dirs = [prefix_info(container, p) for p in prefixes]
+        else:
+            dirs = []
+
+        # Try to fetch a list of files, else return empty
+        if hasattr(resp.output, 'contents'):
```
Comment (tebeka) [inline v3iofs/fs.py:100]: Extract to `_resp_files` utility function.

File: v3iofs/fs.py [hunk 20]
```diff
@@ -37,49 +38,97 @@ class V3ioFS(AbstractFileSystem):
...
+            objects = resp.output.contents
+            files = [object_info(container, o) for o in objects]
+        else:
+            files = []
+
+        pathlist = dirs + files
+
+        if (not hasattr(resp.output, 'common_prefixes')) and (
```
Comment (tebeka) [inline v3iofs/fs.py:108]: Why? What's the case here? Please comment. Also - move to a utility function.

File: v3iofs/fs.py [hunk 21]
```diff
@@ -37,49 +38,97 @@ class V3ioFS(AbstractFileSystem):
...
+                not hasattr(resp.output, 'contents')):
+            dirname, _, filename = path.rpartition("/")
```
Comment (tebeka) [inline v3iofs/fs.py:110]: Add a comment with example of `path -> direname & filename`

File: v3iofs/fs.py [hunk 22]
```diff
@@ -37,49 +38,97 @@ class V3ioFS(AbstractFileSystem):
...
+            resp = self._client.get_container_contents(
+                container=container, path=dirname,
+            )
+            objects = resp.output.contents
+            tempfiles = [object_info(container, o) for o in objects]
+            fullpath = f"/{container}/{path}"
+            for f in tempfiles:
+                if f["name"] == fullpath:
+                    pathlist.append(f)
```
Comment (tebeka) [inline v3iofs/fs.py:119]: We can `break` here, only one file will match

File: v3iofs/fs.py [hunk 23]
```diff
@@ -37,49 +38,97 @@ class V3ioFS(AbstractFileSystem):
...
+
+        if (not hasattr(resp.output, 'common_prefixes')) and (
+                not hasattr(resp.output, 'contents')):
+            dirname, _, filename = path.rpartition("/")
+            resp = self._client.get_container_contents(
+                container=container, path=dirname,
+            )
+            objects = resp.output.contents
+            tempfiles = [object_info(container, o) for o in objects]
+            fullpath = f"/{container}/{path}"
```
Comment (tebeka) [inline v3iofs/fs.py:116]: We construct the original path after we destructed it. Looks like a code smell, maybe we can keep the original path in a variable.

File: v3iofs/fs.py [hunk 24]
```diff
@@ -37,49 +38,97 @@ class V3ioFS(AbstractFileSystem):
...
+            for f in tempfiles:
+                if f["name"] == fullpath:
+                    pathlist.append(f)
+            if not pathlist:
```
Comment (tebeka) [inline v3iofs/fs.py:120]: We don't need this since the same check is done below.

File: v3iofs/fs.py [hunk 25]
```diff
@@ -89,86 +138,111 @@ def copy(self, path1, path2, **kwargs):
...
 def container_path(container):
-    return f'/{container.name}'
+    return f"/{container.name}"
 
 
 def container_info(container):
     return {
-        'name': container.name,
-        'size': None,
-        'created': parse_time(container.creation_date),
```
Comment (tebeka) [inline v3iofs/fs.py:130]: Why did you remove the `created` ?

File: v3iofs/fs.py [hunk 26]
```diff
@@ -89,86 +138,111 @@ def copy(self, path1, path2, **kwargs):
...
+        "name": container.name,
+        "size": None,
     }
 
 
 def prefix_path(container_name, prefix):
     if not isinstance(prefix, str):
-        prefix = prefix.prefix
+        prefix = prefix.prefix.lstrip()
```
Comment (tebeka) [inline v3iofs/fs.py:216]: Was there a case with whitespace on the left?

File: v3iofs/fs.py [hunk 27]
```diff
@@ -89,86 +138,111 @@ def copy(self, path1, path2, **kwargs):
...
     # prefix already have a leading /
```
Comment (tebeka) [inline v3iofs/fs.py:217]: This comment is misleading

File: v3iofs/fs.py [hunk 28]
```diff
@@ -89,86 +138,111 @@ def copy(self, path1, path2, **kwargs):
...
 def info_of(container_name, obj, name_key):
     return {
-        'name': prefix_path(container_name, getattr(obj, name_key)),
-        'size': getattr(obj, 'size', None),
-        'created': parse_time(obj.creating_time),
-        'mtime': parse_time(obj.last_modified),
-        'atime': parse_time(obj.access_time),
-        'mode': int(obj.mode[1:], base=8),  # '040755'
-        'gid': int(obj.gid, 16),
-        'uid': int(obj.uid, 16),
```
Comment (tebeka) [inline v3iofs/fs.py:163]: Why did you remove all of these?

File: v3iofs/fs.py (hunk 2 above)
Comment (hayesgb) [reply v3iofs/fs.py:75]: A user erroneously passing a path such as:  " /xzy.parquet" would return container = True.  This a) seemed more explicit, particularly when looking at the response object that would be returned in the original code, and b) when paired with the updated code changes, returns an explicit error to the user on the issue with their request.

File: .gitignore (hunk 16 above)
Comment (hayesgb) [reply .gitignore]: Autogenerated.  Removed.

File: v3iofs/fs.py [hunk 29]
```diff
@@ -37,49 +38,97 @@ class V3ioFS(AbstractFileSystem):
...
+            )
 
         self._client = Client(
             endpoint=self._v3io_api,
             access_key=self._v3io_access_key,
-            transport_kind='requests',
+            transport_kind="requests",
         )
 
+    def _details(self, contents, **kwargs):
```
[the remark this answers is not in the document]
Comment (hayesgb) [reply v3iofs/fs.py:57]: This is defensive.  As fsspec continues its development, I've had two instances of updates to fsspec adding parameters to other methods that caused breaking changes.

**Comment by tebeka** [comment]
Done in #2
